src/DI/Factory.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+namespace ILIAS\DI;
+
+use ILIAS\DI\Exceptions\InvalidClassException;
+use ILIAS\DI\Exceptions\NoSuchServiceException;
+
+/**
+ * Class Factory
+ * @package ILIAS\DI
+ *
+ * Create instances of classes using type hinting and the dependency injection container.
+ *
+ * @author Oskar Truffer <ot@studer-raimann.ch>
+ */
+class Factory {
```

Comment:
I do understand that you need this kind of factory to initialize the tasks (there are other options, still), but I'm against considering this a general pattern for instantiation. Thus move this to the Background-Tasks-folder, please. The arguments I see against this kind of factory are:
* Folklore says, ReflectionClass is slow. This is not too relevant for the background tasks, but might be a problem when this factory is used in ordinary requests. We should measure the impact of ReflectionClass before  we make this generally available.
* More severe: This pattern confuses "implements XYZ" with "is named XYZ". This is not the same. By announcing a dependency in a constructor, one (optimally) uses interfaces to describe the services an object depends on. This is the "implement XYZ" part. The array-access on the DIC, on the contrary, gives names to certain services. These are optimally only temporary/local when DI is used properly, but ILIAS is not there yet. The problem with this confusion is: the interfaces a service implements only accidentally coincide with its name. There already are services in ILIAS where this is in fact wrong (`$tpl` vs. `ilTemplate`). When announcing a dependency, I should only say: "I need some service implementing interface XYZ", but with this pattern I, in fact, say "I need the service named XYZ". This will prevent some moves in the future, e.g. having two services implementing the same interface.